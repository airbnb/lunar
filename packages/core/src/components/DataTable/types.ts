import React from 'react';
import { SortDirectionType, Table } from 'react-virtualized';
import { WithStylesProps } from '../../composers/withStyles';

export type TableRef = React.RefObject<Table>;

export type RowHeightOptions = string;
export type HeightOptions = RowHeightOptions | undefined;
export type ColumnLabelCase = 'sentence' | 'title' | 'uppercase' | '';

export type SelectedRows = {
  [key: number]: {
    status: string;
    selectedChildren: Set<number>;
  };
};

export type ChangeLog = {
  [key: number]: {
    [key: string]: {
      newVal: any;
    };
  };
};

type EditCallback = (
  row: TableRow,
  key: string,
  newVal: any,
  event: React.SyntheticEvent<EventTarget>,
) => void;

export type OnEdit = (
  row: TableRow,
  key: string,
  newVal: any,
  event: React.SyntheticEvent<EventTarget>,
) => void;

export type HeaderButton = {
  label: string;
  onClick: (selectedRows: SelectedRows) => () => void;
  display?: boolean;
  displayEditMode?: boolean;
};

export type DefaultDataTableProps = keyof DataTableProps;

export interface DataTableProps {
  /** Height of the column header. */
  columnHeaderHeight?: HeightOptions;
  /** Change all column label keys to UPPERCASE or Title Case or Sentence case */
  columnLabelCase?: ColumnLabelCase;
  /** Keys mapped onto custom column label names. */
  columnToLabel?: ColumnToLabel;
  /** Override default width for specific a column's properties. */
  columnMetadata?: ColumnMetadata;
  /** Array of data rows. */
  data?: ParentRow[];
  /** Default callback on all edits. */
  defaultEditCallback?: EditCallback;
  /** Specifies whether or not editMode can be enabled. */
  editable?: boolean;
  /** Callback for any specific key, called on all edits. */
  editCallbacks?: { [key: string]: EditCallback };
  /** When instant edit is disabled, callback that gets trigged on edit application. */
  enactEditsCallback?: (changeLog: ChangeLog) => void;
  /** If enabled, a special column is rendered that allows row to be expanded. */
  expandable?: boolean;
  /** Extra buttons to render in the header during non-edit mode. */
  extraHeaderButtons?: HeaderButton[];
  /** Filter function to handle searching and filtering.. */
  filterData?: (data: IndexedParentRow[]) => IndexedParentRow[];
  /** Height of the entire table. */
  height?: number;
  /** Renders a Done button. Defaults to Cancel and Apply buttons. */
  instantEdit?: boolean;
  /** References row fields to render as columns, infered from data if not specified. */
  keys?: string[];
  /** Propagated as the 'ref' prop to the underlying react-virtualized Table instance. */
  propagateRef?: TableRef;
  /** Custom renderers mapped to column keys. */
  renderers?: Renderers;
  /** Height of table rows, default for table header and column header height. */
  rowHeight?: RowHeightOptions;
  /** If enabled, a special column is rendered to allows rows to be selected. */
  selectable?: boolean;
  /** If enabled, clicking the row triggers the same function as click the selection checkbox. */
  selectOnRowClick?: boolean;
  /**
   * If true, will set Table height to accomodate showing _all_ rows.
   * This effectively _disables_ virtualized row rendering and may have detrimental
+  * performance implications for rendering many rows.
   * */
  showAllRows?: boolean;
  /** If enabled, renders a border between each column. */
  showColumnDividers?: boolean;
  /** If enabled, renders a border between each row. */
  showRowDividers?: boolean;
  /** Overrides sort and hands control to the parent component. */
  sortOverride?: boolean;
  /** sortBy value if override is enabled. */
  sortByOverride?: string;
  /** sortDirection value if override is enabled. */
  sortDirectionOverride?: SortDirectionType;
  /** sortCallback in override is enabled. */
  sortCallback?: (sortBy: string, sortDirection: SortDirectionType) => void;
  /** Label for the table header. */
  tableHeaderLabel?: string;
  /** Height of the table header. */
  tableHeaderHeight?: HeightOptions;
  /** Width of the entire table. */
  width?: number;
  /** If enabled, every other row will appear in grey. */
  zebra?: boolean;
}

export interface GenericRow {
  data: {
    [key: string]: any;
  };
}

/** The row used by React Virtualized. */
export interface TableRow {
  cellData?: void;
  columnData?: void;
  columnIndex: number;
  dataKey: string;
  isScrolling?: boolean;
  parent?: React.ReactNode;
  rowData: ExpandedRow;
  rowIndex: number;
}

export type Status = string;

type ParentMetadata = {
  colSpanKey?: string;
  children?: ChildRow[];
  status?: string;
};

type ChildMetadata = {
  colSpanKey?: string;
  status?: Status;
};

type IndexedParentMetadata = ParentMetadata & {
  originalIndex: number;
  isChild: boolean;
  children: IndexedChildRow[];
};

type IndexedChildMetadata = ChildMetadata & {
  originalIndex: number;
  children?: undefined;
  isChild: boolean;
};

type ExpandedParentMetadata = IndexedParentMetadata & {
  preExpandedIndex: number;
  parentIndex?: undefined;
  parentOriginalIndex?: undefined;
};

type ExpandedChildMetadata = IndexedChildMetadata & {
  preExpandedIndex?: undefined;
  parentIndex: number;
  parentOriginalIndex: number;
};

interface Row {
  data: {
    [key: string]: any;
  };
}

export interface ChildRow extends Row {
  metadata?: ChildMetadata;
}

export interface ParentRow extends Row {
  metadata?: ParentMetadata;
}

export interface IndexedParentRow extends Row {
  metadata: IndexedParentMetadata;
}
export interface IndexedChildRow extends Row {
  metadata: IndexedChildMetadata;
}

export interface ExpandedParentRow extends Row {
  metadata: ExpandedParentMetadata;
}
export interface ExpandedChildRow extends Row {
  metadata: ExpandedChildMetadata;
}

export type IndexedRow = IndexedChildRow | IndexedParentRow;
export type ExpandedRow = ExpandedParentRow | ExpandedChildRow;

export type RowStyles = {
  [key: string]: string;
};

export type ColumnMetadata = {
  [key: string]: {
    [key: string]: number;
  };
};

export type ColumnToLabel = {
  [key: string]: any;
};

export type RendererProps = {
  /** Row including row data and metadata. */
  row: TableRow;
  /** Key being rendered. */
  key: string;
  /** Whether or not edit mode is enabled. */
  editMode: boolean;
  /** Callback to trigger on cell edit. */
  onEdit: OnEdit;
  /** Whether or not zebra mode is enabled. */
  zebra: boolean;
  /** Theme from Lunar. */
  theme: WithStylesProps['theme'];
};

type Renderer = (props: RendererProps) => NonNullable<React.ReactNode>;

export type Renderers = {
  [key: string]: Renderer;
};

export type WidthProperties = {
  [property: string]: number;
};
