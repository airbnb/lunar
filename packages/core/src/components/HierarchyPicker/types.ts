import { Item as AutocompleteItem } from '../Autocomplete';

export interface ItemShape {
  /** Identifier used in chosen definition. */
  name: string;
  /** Localized content displayed in lieu of name. */
  label?: string;
  /** Long-worded localized description of an item. */
  description?: string;
  /** Optional flag to signal item should not be pickable. */
  readonly?: boolean;
  /** Optional keywords for use in search. */
  keywords?: string;
  /** Optional recursive sub-items. */
  items?: ItemShape[] | null;
  /** Optional menu section label. */
  section?: string;
}

export type FuseMatch = {
  indices?: number[][];
  key: string;
  value: string;
};

export type TreePath = string[];

export interface SearchItemShape extends ItemShape {
  definition: TreePath;
  label: string;
  formattedParents: string;
}

export interface SearchItemResult extends AutocompleteItem {
  item: SearchItemShape;
  matches: FuseMatch[];
}

export type TopicOriginKey = 'Hierarchy' | 'Search';

export type ChoiceDetails = { origin: TopicOriginKey; charCount?: number };

export type Labeler = (chosen: TreePath) => string;

export type Formatter = (chosen: TreePath) => string;

export type ItemRenderer = (
  item: ItemShape,
  selected: boolean,
  focused: boolean,
) => NonNullable<React.ReactNode>;

export type ItemPickedHandler = (chosen: TreePath | null, details?: ChoiceDetails) => void;

export type DeepFocusHandler = () => void;

export type ToggleHandler = (open: boolean) => void;

export type ShallowFocusHandler = () => void;

export type SubTreeHandler = (
  path: TreePath,
  onDeeper?: DeepFocusHandler,
  immediate?: boolean,
) => void;
