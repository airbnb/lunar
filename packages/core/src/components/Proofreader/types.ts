/* eslint-disable camelcase */

export type ProofreadRuleMatch = {
  message?: string;
  short_message?: string;
  offset?: number;
  length?: number;
  found?: string;
  replacements?: Array<string>;
  rule_id?: string;
};

export type ProofreaderParams = {
  text: string;
  locale: string;
  action: 'check';
};

export type ProofreaderResponse = {
  proofread: {
    matches: ProofreadRuleMatch[];
  };
};

export type DefinitionShape = {
  locale: string;
  label: string;
};

export type ExtraProofreadProps = {
  isRuleHighlighted?: (rule: ProofreadRuleMatch) => boolean;
  isRuleSecondary?: (rule: ProofreadRuleMatch) => boolean;
};
