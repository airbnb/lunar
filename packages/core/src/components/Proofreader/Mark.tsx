import React, { useRef, useEffect, useCallback } from 'react';
import useStyles from '../../hooks/useStyles';
import { ProofreadRuleMatch } from './types';
import { markStyleSheet } from './styles';

export type MarkProps = {
  children: string;
  error: ProofreadRuleMatch;
  highlighted?: boolean;
  secondary?: boolean;
  selected?: boolean;
  shadow?: boolean;
  onSelect: (error: ProofreadRuleMatch, top: number, left: number) => void;
};

export default function Mark({
  children,
  error,
  highlighted,
  secondary,
  selected,
  shadow = false,
  onSelect,
}: MarkProps) {
  const [styles, cx] = useStyles(markStyleSheet);
  const ref = useRef<HTMLSpanElement | null>(null);
  const active = highlighted || selected;

  const handleSelect = useCallback(() => {
    if (ref.current) {
      onSelect(error, ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
    }
  }, [error, onSelect]);

  // When being used as a shadow, the mark cannot be clicked.
  // So we automatically fire the handler when selected state changes.
  useEffect(() => {
    if (shadow && selected) {
      handleSelect();
    }
  }, [shadow, selected, handleSelect]);

  return (
    <button className={cx(styles.button)} type="button" onClick={handleSelect}>
      <mark
        ref={ref}
        className={cx(
          styles.mark,
          active && styles.mark_highlight,
          secondary && styles.markSecondary,
          secondary && active && styles.markSecondary_highlight,
        )}
      >
        {children}
      </mark>
    </button>
  );
}
