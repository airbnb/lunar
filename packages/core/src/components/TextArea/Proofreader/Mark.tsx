import React from 'react';
import withStyles, { WithStylesProps } from '../../../composers/withStyles';

export type Props = {
  children: NonNullable<React.ReactNode>;
  selected: boolean;
  onSelect: (top: number, left: number) => void;
};

class Mark extends React.PureComponent<Props & WithStylesProps> {
  ref = React.createRef<HTMLSpanElement>();

  componentDidMount() {
    this.handleSelect();
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.selected !== prevProps.selected) {
      this.handleSelect();
    }
  }

  private handleSelect = () => {
    const { ref } = this;

    if (this.props.selected && ref.current) {
      this.props.onSelect(ref.current.offsetTop + ref.current.offsetHeight, ref.current.offsetLeft);
    }
  };

  render() {
    const { cx, children, selected, styles } = this.props;

    return (
      <mark className={cx(styles.mark, selected && styles.mark_selected)} ref={this.ref}>
        {children}
      </mark>
    );
  }
}

export default withStyles(({ color, ui }) => ({
  mark: {
    position: 'relative',
    color: 'transparent',
    backgroundColor: 'transparent',
    opacity: 0.75,
    padding: 1,
    margin: -1,
    transition: 'opacity .2s, background .2s',

    '::after': {
      position: 'absolute',
      display: 'block',
      content: '""',
      bottom: 0,
      left: 0,
      right: 0,
      height: 2,
      borderRadius: ui.borderRadius,
      backgroundColor: color.core.danger[2],
    },
  },

  mark_selected: {
    opacity: ui.disabledOpacity,
    borderTopLeftRadius: ui.borderRadius,
    borderTopRightRadius: ui.borderRadius,
    backgroundColor: color.core.danger[2],
  },
}))(Mark);
