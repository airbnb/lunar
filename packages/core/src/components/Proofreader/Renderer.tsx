import React from 'react';
import Interweave from '../Interweave';
import Mark, { MarkProps } from './Mark';
import { ProofreadRuleMatch } from './types';

export type RendererProps = {
  errors: ProofreadRuleMatch[];
  isRuleHighlighted?: (rule: ProofreadRuleMatch) => boolean;
  isRuleSecondary?: (rule: ProofreadRuleMatch) => boolean;
  onSelectError: MarkProps['onSelect'];
  selectedError?: ProofreadRuleMatch | null;
  shadow?: boolean;
  value: string;
};

export default function Renderer({
  errors,
  isRuleHighlighted,
  isRuleSecondary,
  onSelectError,
  selectedError,
  shadow = false,
  value,
}: RendererProps) {
  // If no errors, return the value without marks
  if (errors.length === 0) {
    return (
      <div>
        <Interweave disableMatchers content={value} />
      </div>
    );
  }

  // Sort errors by offset otherwise slicing does not work
  errors.sort((a, b) => (a.offset ?? 0) - (b.offset ?? 0));

  const content: NonNullable<React.ReactNode>[] = [];
  let lastIndex = 0;

  errors.forEach((error) => {
    const offset = error.offset ?? 0;
    const length = error.length ?? 0;

    if (offset > value.length) {
      return;
    }

    // Extract previous string
    content.push(
      <Interweave
        key={`${lastIndex}-${offset}`}
        disableMatchers
        content={value.slice(lastIndex, offset)}
      />,
    );

    // Set new last index
    lastIndex = offset + length;

    // Extract match and wrap in a component
    const word = value.slice(offset, lastIndex);

    content.push(
      <Mark
        key={`${word}-${offset}`}
        error={error}
        highlighted={isRuleHighlighted?.(error)}
        secondary={isRuleSecondary?.(error)}
        selected={error === selectedError}
        shadow={shadow}
        onSelect={onSelectError}
      >
        {word}
      </Mark>,
    );
  });

  // Extract any remaining value
  const final = value.slice(lastIndex);

  content.push(
    <Interweave key={`${lastIndex}-${lastIndex + final.length}`} disableMatchers content={final} />,
  );

  // Add a fake character to the end of the text when a shadow. This solves a handful of bugs
  // in which trailing new lines in combination with scroll position do not work correctly.
  if (shadow) {
    content.push('.');
  }

  return <div>{content}</div>;
}
