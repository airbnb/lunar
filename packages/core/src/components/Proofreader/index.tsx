import React from 'react';
import debounce from 'lodash/debounce';
import BaseTextArea from '../private/BaseTextArea';
import Dropdown from '../Dropdown';
import withStyles, { WithStylesProps } from '../../composers/withStyles';
import { NO_LOCALE, NON_WORD_REGEX, ARROW_KEYS } from './constants';
import Renderer from './Renderer';
import ErrorMenu from './ErrorMenu';
import ControlBar from './ControlBar';
import {
  ProofreadRuleMatch,
  ProofreaderResponse,
  ExtraProofreadProps,
  ProofreaderParams,
} from './types';
import { FormInputProps } from '../private/FormInput';
import { styleSheet } from './styles';
import { selectAppropriateLocale, checkForAirbnbErrors } from './helpers';

function defaultIsRuleHighlighted(rule: ProofreadRuleMatch) {
  return false;
}

function defaultIsRuleSecondary(rule: ProofreadRuleMatch) {
  return false;
}

export type Position = {
  left: number;
  top: number;
};

export type ProofreaderProps = Pick<FormInputProps, 'important' | 'noTranslate'> &
  ExtraProofreadProps & {
    id: string;
    locale?: string;
    name: string;
    onChange: (value: string, event: React.ChangeEvent<HTMLTextAreaElement>) => void;
    onCheckText: (params: ProofreaderParams) => Promise<ProofreaderResponse>;
    value: string;
  };

export type ProofreaderState = {
  errors: ProofreadRuleMatch[];
  loading: boolean;
  position: Position | null;
  selectedError: ProofreadRuleMatch | null;
  selectedLocale: string | null;
  text: string;
};

export type Snapshot = {
  selectionStart?: number;
  selectionEnd?: number;
};

export class Proofreader extends React.Component<
  ProofreaderProps & WithStylesProps,
  ProofreaderState,
  Snapshot
> {
  static defaultProps = {
    isRuleHighlighted: defaultIsRuleHighlighted,
    isRuleSecondary: defaultIsRuleSecondary,
    locale: NO_LOCALE,
    value: '',
  };

  caretRef = React.createRef<HTMLDivElement>();

  controlsRef = React.createRef<HTMLDivElement>();

  shadowRef = React.createRef<HTMLDivElement>();

  textareaRef = React.createRef<HTMLTextAreaElement>();

  inputRaf: number = 0;

  syncScrollRaf: number = 0;

  scrollTimeout: number = 0;

  state: ProofreaderState = {
    errors: [],
    loading: false,
    position: null,
    selectedError: null,
    selectedLocale: null,
    text: this.props.value,
  };

  getSnapshotBeforeUpdate() {
    if (!this.textareaRef.current) {
      return null;
    }

    return {
      selectionStart: this.textareaRef.current.selectionStart,
      selectionEnd: this.textareaRef.current.selectionEnd,
    };
  }

  componentDidMount() {
    this.selectAppropriateLocale();
  }

  componentDidUpdate(prevProps: ProofreaderProps, prevState: ProofreaderState, snapshot: Snapshot) {
    const { locale, value } = this.props;

    if (locale !== prevProps.locale) {
      this.selectAppropriateLocale();
    }

    if (value === prevProps.value) {
      // if value hasn't changed but other props have, make sure to retain original cursor position
      this.setCursorPosition(snapshot);
    } else {
      this.setState(
        {
          text: value,
          errors: [],
          selectedError: null,
        },
        () => {
          // reset cursor position to original position after updating value
          this.setCursorPosition(snapshot);
        },
      );

      // Start checking when we encounter a non-word character so we can avoid
      // reporting misspellings for half-typed words.
      if (value.length > 0 && value.slice(-1).match(NON_WORD_REGEX)) {
        this.checkTextDebounced();
      }

      // If the user backspaces and removes a chunk of text, the scroll position
      // and dimensions may be out of sync, so we need to manually update them.
      // Don't use RAF though as it may take too long to render.
      if (value.length < prevProps.value.length) {
        window.setTimeout(this.performSyncScroll, 0);
      }
    }
  }

  componentWillUnmount() {
    if (this.syncScrollRaf) {
      cancelAnimationFrame(this.syncScrollRaf);
    }

    if (this.inputRaf) {
      cancelAnimationFrame(this.inputRaf);
    }
  }

  setCursorPosition(snapshot: Snapshot) {
    if (!snapshot) return;

    const { selectionStart, selectionEnd } = snapshot;
    const textarea = this.textareaRef.current;

    if (textarea && selectionStart && selectionEnd) {
      textarea.setSelectionRange(selectionStart, selectionEnd);
    }
  }

  checkText(): Promise<void> {
    const { text, selectedLocale } = this.state;

    this.setState({
      loading: true,
      position: null,
      selectedError: null,
    });

    // Check for Airbnb misspellings
    const customErrors = checkForAirbnbErrors(text);

    if (!text || !selectedLocale || selectedLocale === NO_LOCALE) {
      this.setState({
        errors: [...customErrors],
        loading: false,
      });

      return Promise.resolve();
    }

    // Check with language tool
    const params: ProofreaderParams = {
      text,
      locale: selectedLocale,
      action: 'check',
    };

    return this.props
      .onCheckText(params)
      .then(({ proofread }: ProofreaderResponse) => {
        this.setState({
          errors: [...customErrors, ...proofread.matches],
          loading: false,
        });
      })
      .catch(() => {
        // Swallow errors
        this.setState({
          errors: [...customErrors],
          loading: false,
        });
      });
  }

  checkTextAndClearErrors() {
    if (this.state.errors.length > 0) {
      this.setState({
        errors: [],
      });
    }

    this.checkTextDebounced();
  }

  checkTextDebounced = debounce(this.checkText, 2000, {
    leading: true,
  });

  getCaretOffset(): Position {
    let top = 0;
    let left = 0;

    if (this.caretRef.current) {
      const child = this.caretRef.current.children[1] as HTMLElement;

      top = child.offsetTop;
      left = child.offsetLeft;
    }

    return { top, left };
  }

  selectAppropriateLocale() {
    this.setState(
      {
        selectedLocale: selectAppropriateLocale(this.props.locale).selectedLocale,
      },
      () => {
        this.checkTextAndClearErrors();
      },
    );
  }

  /**
   * When clicking into a textarea, or moving the cursor around with arrow keys,
   * attempt to detect any current errors at the defined offsets.
   */
  selectErrorAtPosition(position: number) {
    const { errors } = this.state;

    if (errors.length === 0) {
      return;
    }

    const selectedError =
      errors.find(
        (error) =>
          position >= (error.offset ?? 0) && position <= (error.offset ?? 0) + (error.length ?? 0),
      ) ?? null;

    this.setState({
      selectedError,
    });
  }

  syncScrollPositions = () => {
    if (!this.syncScrollRaf) {
      this.syncScrollRaf = requestAnimationFrame(this.performSyncScroll);
    }
  };

  performSyncScroll = () => {
    this.syncScrollRaf = 0;

    if (this.shadowRef.current && this.textareaRef.current) {
      const { current: shadow } = this.shadowRef;
      const { current: textarea } = this.textareaRef;

      shadow.style.minHeight = textarea.style.minHeight;
      shadow.style.height = textarea.style.height;
      shadow.scrollTop = textarea.scrollTop;
    }
  };

  private handleTextAreaClick = (event: React.MouseEvent<HTMLTextAreaElement>) => {
    this.selectErrorAtPosition(event.currentTarget.selectionStart);
  };

  private handleTextAreaKeyDown = (event: React.KeyboardEvent<HTMLTextAreaElement>) => {
    if (ARROW_KEYS.includes(event.key)) {
      this.selectErrorAtPosition(event.currentTarget.selectionStart);
    }
  };

  /**
   * When a user is typing, we need to capture the offset of the caret within the textarea,
   * and determine when they are typing near the bottom of the container. If so,
   * we need to set the scroll top to max so text doesn't go behind the control bar.
   */
  private handleInput = () => {
    if (!this.inputRaf) {
      this.inputRaf = requestAnimationFrame(this.handleInputScroll);
    }
  };

  private handleInputScroll = () => {
    this.inputRaf = 0;

    if (!this.textareaRef.current || !this.controlsRef.current) {
      return;
    }

    const el = this.textareaRef.current;
    const offset = this.getCaretOffset();
    const scrollBuffer = this.controlsRef.current.offsetHeight + 35;

    // At the bottom
    if (offset.top >= el.scrollHeight - scrollBuffer) {
      el.scrollTop = el.scrollHeight;
      // In the middle of the textarea, below the scroll
    } else if (offset.top >= el.offsetHeight + el.scrollTop - scrollBuffer) {
      el.scrollTop += scrollBuffer;
    }
  };

  /**
   * When a marked/highlighted word is focused, open the error menu with the marks position.
   */
  private handleOpenErrorMenu = (error: ProofreadRuleMatch, top: number, left: number) => {
    this.setState({
      selectedError: error,
      position: {
        top: top - (this.textareaRef.current ? this.textareaRef.current.scrollTop : 0),
        left,
      },
    });
  };

  /**
   * When a replacement word is clicked within the error menu, replace the current text with the
   * new word at the define error offsets.
   */
  private handleReplaceText = (error: ProofreadRuleMatch, replacement: string) => {
    const { offset = 0 } = error;
    const { errors, text } = this.state;
    let newText = '';

    newText += text.slice(0, offset);
    newText += replacement;
    newText += text.slice(offset + error.length!);

    this.setState(
      {
        text: newText,
        // Filter out this error
        errors: errors.filter((e) => e !== error),
        // Close the error menu
        position: null,
        selectedError: null,
      },
      () => {
        // Kick off a non-debounced checkText, as this is coming directly from a replacement:
        this.checkText();

        const newOffset = offset + replacement.length;
        const textarea = this.textareaRef.current;

        if (!textarea) {
          return;
        }

        // Focus the textarea so we can reset the cursor position.
        textarea.focus();
        textarea.setSelectionRange(newOffset, newOffset);

        // Blurring and focusing the textarea will scroll the textarea to the cursor position,
        // which is what we need to do to avoid all our position syncing shenanigans.
        textarea.blur();
        textarea.focus();
      },
    );

    if (this.props.onChange) {
      // We don't have a real event here, so fake one
      const event = {
        nativeEvent: new Event('change'),
        currentTarget: this.textareaRef,
        target: this.textareaRef,
        preventDefault() {},
      };

      this.props.onChange(
        newText,
        // @ts-ignore
        event,
      );
    }
  };

  /**
   * Keep shadow scroll position in sync with the textarea.
   */
  private handleScroll = () => {
    this.syncScrollPositions();

    if (this.state.position) {
      this.setState({
        selectedError: null,
        position: null,
      });
    }

    // When the textarea has trailing newlines, the scroll event
    // will not fire when scrolled to the very bottom, so we
    // get into a weird state where the scroll positions don't match.
    // Add a small timeout to fix this issue.
    window.clearTimeout(this.scrollTimeout);

    this.scrollTimeout = window.setTimeout(this.syncScrollPositions, 500);
  };

  private handleSelectLocale = (locale: string) => {
    this.setState(
      {
        selectedLocale: locale,
      },
      () => {
        this.checkTextAndClearErrors();
      },
    );
  };

  render() {
    const {
      children,
      cx,
      important,
      isRuleHighlighted,
      isRuleSecondary,
      localeMenuMaxHeight,
      onChange,
      onCheckText,
      styles,
      theme,
      ...props
    } = this.props;
    const { position, errors, loading, selectedError, selectedLocale, text } = this.state;
    const caretPosition =
      (this.textareaRef.current && this.textareaRef.current.selectionStart) || 0;
    const highlightsProps = {
      className: cx(styles.highlights, important && styles.highlights_important),
      ref: this.shadowRef,
    };

    if (this.props.noTranslate) {
      highlightsProps.className += ' notranslate';
    }

    return (
      <div className={cx(styles.proofread)}>
        {/* Shadow text for displaying underlined words. */}
        <div {...highlightsProps}>
          <Renderer
            shadow
            value={text}
            errors={errors}
            selectedError={selectedError}
            isRuleHighlighted={isRuleHighlighted}
            isRuleSecondary={isRuleSecondary}
            onSelectError={this.handleOpenErrorMenu}
          />
        </div>

        {/* Track the top/left offset of the caret within the textarea. */}
        {caretPosition > 0 && text && (
          <div ref={this.caretRef} className={cx(styles.caret)}>
            <span>{text.slice(0, caretPosition)}</span>
            <span>{text.slice(caretPosition)}.</span>
          </div>
        )}

        <BaseTextArea
          {...props}
          spellCheck={false}
          value={this.state.text}
          propagateRef={this.textareaRef}
          important={important}
          onChange={onChange}
          onClick={this.handleTextAreaClick}
          onKeyDown={this.handleTextAreaKeyDown}
          onScroll={this.handleScroll}
          onInput={this.handleInput}
        />

        {position && selectedError && (
          <Dropdown {...position} zIndex={5}>
            <ErrorMenu error={selectedError} onReplaceText={this.handleReplaceText} />
          </Dropdown>
        )}

        <div
          ref={this.controlsRef}
          className={cx(styles.controls, important && styles.controls_important)}
        >
          <ControlBar
            loading={loading}
            locale={selectedLocale!}
            localeMenuMaxHeight={localeMenuMaxHeight}
            errors={errors}
            onSelectLocale={this.handleSelectLocale}
          />
        </div>
      </div>
    );
  }
}

export default withStyles(styleSheet, {
  passThemeProp: true,
})(Proofreader);
