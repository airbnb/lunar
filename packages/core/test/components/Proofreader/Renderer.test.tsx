import React from 'react';
import { render } from 'rut-dom';
import Renderer, { RendererProps } from '../../../src/components/Proofreader/Renderer';
import { ProofreadRuleMatch } from '../../../src/components/Proofreader/types';
import Mark from '../../../src/components/Proofreader/Mark';

describe('<Renderer />', () => {
  const props: RendererProps = {
    errors: [],
    value: 'This is a really long input string.',
    onSelectError() {},
  };

  const error: ProofreadRuleMatch = {
    found: 'really long',
    length: 11,
    message: 'Invalid',
    offset: 10,
    replacements: [],
    rule_id: 'TEST',
  };

  const debugOptions = {
    log: false,
    reactElements: false,
  };

  it('renders full string if no errors', () => {
    const { root } = render<RendererProps>(<Renderer {...props} />);

    expect(root.findAt('div', 'first').debug(debugOptions)).toMatchSnapshot();
  });

  it('wraps errors with marks', () => {
    const { root } = render<RendererProps>(
      <Renderer
        {...props}
        errors={[
          error,
          {
            found: 'string',
            length: 6,
            message: 'Invalid',
            offset: 28,
            replacements: [],
            rule_id: 'WORD',
          },
        ]}
      />,
    );

    expect(root.findAt('div', 'first').debug(debugOptions)).toMatchSnapshot();
  });

  it('can mark as secondary', () => {
    const { root } = render<RendererProps>(
      <Renderer {...props} errors={[error]} isRuleSecondary={(rule) => rule.rule_id === 'TEST'} />,
    );

    expect(root.findOne(Mark)).toHaveProp('secondary', true);
  });

  it('can mark as highlighted', () => {
    const { root } = render<RendererProps>(
      <Renderer
        {...props}
        errors={[error]}
        isRuleHighlighted={(rule) => rule.rule_id === 'TEST'}
      />,
    );

    expect(root.findOne(Mark)).toHaveProp('highlighted', true);
  });

  it('marks as selected if errors match', () => {
    const { root } = render<RendererProps>(
      <Renderer {...props} errors={[error]} selectedError={error} />,
    );

    expect(root.findOne(Mark)).toHaveProp('selected', true);
  });

  it('calls `onSelectError` when a mark is clicked', () => {
    const spy = jest.fn();
    const { root } = render<RendererProps>(
      <Renderer {...props} errors={[error]} onSelectError={spy} />,
      {
        mockRef: () => ({ offsetTop: 5, offsetRight: 10, offsetLeft: 15, offsetHeight: 100 }),
      },
    );

    root.findOne('button').dispatch('onClick');

    expect(spy).toHaveBeenCalledWith(error, 105, 15);
  });
});
