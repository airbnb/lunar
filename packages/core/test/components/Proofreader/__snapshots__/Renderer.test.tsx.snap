// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Renderer /> renders full string if no errors 1`] = `
"<div>
  <span>This is a really long input string.</span>
</div>"
`;

exports[`<Renderer /> wraps errors with marks 1`] = `
"<div>
  <span>This is a </span>
  <button className=\\"button\\" type=\\"button\\" onClick={func()}>
    <mark className=\\"mark\\">
      really long
    </mark>
  </button>
  <span> input </span>
  <button className=\\"button\\" type=\\"button\\" onClick={func()}>
    <mark className=\\"mark\\">
      string
    </mark>
  </button>
  <span>.</span>
</div>"
`;
